include/verilated_vpi.cpp

```diff
@@ -2812,13 +2824,1086 @@ vpiHandle vpi_put_value(vpiHandle object, p_vpi_value valuep, p_vpi_time /*time_
+                }
+            } else {
+                for (int i = 0; i < num; i++) {
+                    for (int j = 0; j < words; j++) {
+                        rawvals[i * words + j] = ptr[index * words + j];
+                    }
+
+                    index = (--index + size) % size;
+                }
+            }
+        }
+
+        return;
+    } else if (arrayvalue_p->format == vpiRealVal) {
+        double *reals;
+
+        if (arrayvalue_p->flags & vpiUserAllocFlag) {
+            reals = arrayvalue_p->value.reals;
+        } else {
+            if (VL_UNCOVERABLE(num >= VL_VALUE_STRING_MAX_WORDS)) {
+                VL_FATAL_MT(__FILE__, __LINE__, "",
+                            "vpi_get_value_array with more than VL_VALUE_STRING_MAX_WORDS; "
+                            "increase and recompile");
+            }
+
+            reals = (double*)out_ptr;
+            arrayvalue_p->value.reals = reals;
+        }
+
+        if (varp->vltype() == VLVT_UINT64) {
+            double *ptr = reinterpret_cast<double*>(vop->varDatap());
+
+            if (vop->rangep()->left() < vop->rangep()->right()) {
+                for (int i = 0; i < num; i++) {
+                    reals[i] = ptr[index];
+                    index = ++index % size;
+                }
+            } else {
+                for (int i = 0; i < num; i++) {
+                    reals[i] = ptr[index];
+                    index = (--index + size) % size;
+                }
+            }
+        }
+
+        return;
+    }
+
+    arrayvalue_p->value.rawvals = nullptr;
+
+    VL_VPI_ERROR_(__FILE__, __LINE__, "%s: Unsupported format (%s) as requested for %s", __func__,
+                  VerilatedVpiError::strFromVpiVal(arrayvalue_p->format), vop->fullname());
+}
+
+void vpi_get_value_array(vpiHandle object, p_vpi_arrayvalue arrayvalue_p,
+                         PLI_INT32* index_p, PLI_UINT32 num) {
+    VL_DEBUG_IF_PLI(VL_DBG_MSGF("- vpi: vpi_get_value_array %p\n", object););
+    VerilatedVpiImp::assertOneCheck();
+
+    VL_VPI_ERROR_RESET_();
+    if (VL_UNLIKELY(!object)) return;
+
+    const VerilatedVpioVar* const vop = VerilatedVpioVar::castp(object);
+    if (VL_UNLIKELY(!vop)) {
+        VL_VPI_ERROR_(__FILE__, __LINE__, "%s: Unsupported vpiHandle (%p)", __func__, object);
+    }
+
+    if (vop->type() != vpiRegArray) {
+        VL_VPI_ERROR_(__FILE__, __LINE__, "%s: Unsupported type (%p, %s)", __func__,
+            object, VerilatedVpiError::strFromVpiObjType(vop->type()));
+    }
+
+    const VerilatedVar* const varp = vop->varp();
+
+    if (varp->dims() != 2) {
+        VL_VPI_ERROR_(__FILE__, __LINE__, "%s: object %s has unsupported number of indices (%d)",
+            __func__, vop->fullname(), varp->dims());
+    }
+
+    if (VL_UNLIKELY(!vop->rangep())) return;
+
+    int minRange = fmin(vop->rangep()->left(), vop->rangep()->right());
+    int maxRange = fmax(vop->rangep()->left(), vop->rangep()->right());
```

Comment:
```suggestion
    const int minRange = std::min(vop->rangep()->left(), vop->rangep()->right());
    const int maxRange = std::max(vop->rangep()->left(), vop->rangep()->right());
```

Check other places too for missing const.